File: day03/day03.py
#!/usr/bin/env python3

import os
import re
import time
from typing import List

SCRIPT_DIR = os.path.dirname(os.path.realpath(__file__))


def read_input(path: str) -> str:
    with open(path, 'r') as f:
        lines = f.read()
    return lines


# =========================================

def run1(data: str):
    pattern = r"mul\((\d+),(\d+)\)"
    matches = re.findall(pattern, data)
    return sum([int(m[0]) * int(m[1]) for m in matches])

# =========================================

def run2(data: str):
    pattern = r"(?:mul\((\d+),(\d+)\)|(do\(\))|(don't\(\)))"
    matches = re.findall(pattern, data)
    enabled = True
    result = 0
    for m in matches:
        match m:
            case ('', '', 'do()', ''):
                enabled = True
            case ('', '', '', "don't()"):
                enabled = False
            case _:
                if enabled:
                    result += int(m[0]) * int(m[1])
    return result


# INPUT = f"{SCRIPT_DIR}/input_test.txt"
INPUT = f"{SCRIPT_DIR}/input.txt"
data = read_input(INPUT)

start1 = time.time()
result1 = run1(data)
duration1 = (time.time() - start1) * 1000
print("Result1 = ", result1, f"    \t(in {duration1:.6f}ms)")

start2 = time.time()
result2 = run2(data)
duration2 = (time.time() - start2) * 1000
print("Result2 = ", result2, f"    \t(in {duration2:.6f}ms)")

File: day03/day03.rb
#!/usr/bin/env ruby

def step1(data)
    matches = data.scan(/mul\((\d+),(\d+)\)/).map {|x| [x[0].to_i, x[1].to_i]}
    return matches.map{|x| x[0]*x[1]}.sum
end

def step2(data)
    matches = data.scan(/(?:mul\((\d+),(\d+)\)|(do\(\))|(don't\(\)))/)
    result = 0
    disabled = false
    matches.each do |m|
      disabled = false if m[2] == "do()"
      disabled = true if m[3] == "don't()"
      result += m[0].to_i * m[1].to_i if m[0] != nil and ! disabled
    end
    return result
end


if __FILE__ == $0

    INPUT_FILE = "input.txt"
    data = File.read(File.join(__dir__, INPUT_FILE))

    start = Time.now
    result1 = step1(data)
    finish = Time.now
    puts "Result1 = #{result1}   \t(in #{'%.6f' % ((finish - start) * 1000)}ms)"

    start = Time.now
    result2 = step2(data)
    finish = Time.now
    puts "Result2 = #{result2}   \t(in #{'%.6f' % ((finish - start) * 1000)}ms)"

end
